ledger-tool/src/main.rs

```diff
@@ -28,13 +36,198 @@ use std::{
+    txn_sig: Option<String>,
+    fee: Option<f64>,
+    num_instructions: Option<usize>,
+    num_accounts: Option<usize>,
+    account_balance_infos: Vec<AccountBalanceInfo>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq)]
+struct AccountBalanceInfo {
+    account_pubkey: String,
+    account_pre_balance: f64,
+    account_post_balance: f64,
+}
+
+fn output_slot_to_csv(
+    blockstore: &Blockstore,
+    slot: Slot,
+    txn_wtr: &mut Writer<File>,
+    instruction_wtr: &mut Writer<File>,
+    ledger_path: &PathBuf,
+) -> Result<(), String> {
+    let entries = blockstore
+        .get_slot_entries(slot, 0, None)
+        .map_err(|err| format!("Failed to load entries for slot {}: {}", slot, err))?;
+
+    for entry in entries {
+        for transaction in entry.transactions {
+            // Skip any vote transactions
+            let program_pubkey = transaction
+                .message
+                .account_keys[transaction.message.instructions[0].program_id_index as usize];
+            if program_pubkey == solana_vote_program::id() { continue; }
+
+            let transaction_status: Option<RpcTransactionStatusMeta> = blockstore
+                .read_transaction_status((transaction.signatures[0], slot))
+                .unwrap_or_else(|err| {
+                    eprintln!(
+                        "Failed to read transaction status for {} at slot {}: {}",
+                        transaction.signatures[0], slot, err
+                    );
+                    None
+                })
+                .map(|transaction_status| transaction_status.into());
+
+            for instruction in &transaction.message.instructions {
+                let instruction_info = build_instruction_info(
+                    &instruction,
+                    &transaction,
+                    &transaction_status
+                );
+                instruction_wtr.serialize(&instruction_info);
+            }
+
+            let txn_info = build_txn_info(
+                slot,
+                &ledger_path,
+                &transaction,
+                &transaction_status
+            );
+            txn_wtr.serialize(&txn_info);
+        }
+    }
+    instruction_wtr.flush();
+    txn_wtr.flush();
+    Ok(())
+}
+
+fn build_instruction_info(
+    instruction: &CompiledInstruction,
+    transaction: &Transaction,
+    transaction_status: &Option<RpcTransactionStatusMeta>
+) -> InstructionInfo {
+    let program_pubkey = transaction
+        .message
+        .account_keys[instruction.program_id_index as usize];
+
+    let mut instruction_info = InstructionInfo{
+        txn_sig: Some(transaction.signatures[0].to_string()),
+        program_pubkey: Some(bs58::encode(program_pubkey).into_string()),
+        ..Default::default()
+    };
+
+    // Need to figure out how to decode program instruction to String
+    instruction_info.program_instruction = Some("foo".to_string());
+//    instruction_info.program_instruction =
```

Comment:
@mvines I'm trying to reproduce the human readable output from here (https://github.com/solana-labs/solana/blob/master/cli/src/display.rs#L108-L115) . The CLI writer is writing to String::utf8, but when I try to do str::from_utf8 on the instruction data here, it panics.  Is there a cleaner way to print to string the instruction.data?

Reply:
Nevermind, figure it out.